torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -65,228 +52,65 @@ class DataLoader {
-      }
-    }
-  }
-
-  optional<BatchRequest> get_batch_request() {
-    auto indices = sampler_.next(options_.batch_size);
-    if (!indices ||
-        (indices->size() < options_.batch_size && options_.drop_last)) {
-      return nullopt;
-    }
-    AT_ASSERT(indices->size() > 0);
-    return indices;
-  }
-
-  template <typename T>
-  void push_job(T value) {
-    shuttle_.push_job({std::move(value), sequence_number_++});
-  }
-
-  std::unique_ptr<detail::sequencers::Sequencer<Result>> new_sequencer() {
-    if (options_.enforce_ordering) {
-      return torch::make_unique<detail::sequencers::OrderedSequencer<Result>>(
-          options_.max_jobs);
-    }
-    return torch::make_unique<detail::sequencers::NoSequencer<Result>>();
-  }
-
-  /// The options the `DataLoader` was configured with.
-  const FullDataLoaderOptions options_;
-
-  /// The dataset for the main thread, only has a value if the number of
-  /// worker threads was configured as zero, meaning the main thread has to do
-  /// all the work (synchronously). NOTE: Really want this to be on the heap
-  /// when empty, therefore `unique_ptr` and not `optional`.
-  std::unique_ptr<Dataset> main_thread_dataset_;
-
-  /// The sampler with which new batch requests are created.
-  Sampler sampler_;
-
-  /// The sequence number for the *next* batch to be retrieved from the
-  /// dataset.
-  size_t sequence_number_ = 0;
-
-  /// The worker threads, running the `worker_thread()` method.
-  std::vector<std::thread> workers_;
-
-  /// The `DataShuttle` which takes care of the life cycle of a job.
-  detail::DataShuttle<Job, Result> shuttle_;
+  Impl impl_;
+}; // namespace data
 
-  /// The `Sequencer`, which handles optional ordering of batches.
-  std::unique_ptr<detail::sequencers::Sequencer<Result>> sequencer_;
+template <typename Dataset, typename Sampler>
+using StatelessDataLoader =
+    DataLoader<detail::dataloader::StatelessImpl<Dataset, Sampler>>;
 
-  /// True if the `DataLoader` has joined its worker threads.
-  bool joined_ = false;
-}; // namespace data
+template <typename Dataset>
+using StatefulDataLoader =
+    DataLoader<detail::dataloader::StatefulImpl<Dataset>>;
 
-/// Creates a new `DataLoader`, inferring the necessary template types from
-/// the given arguments.
 template <typename Dataset, typename Sampler>
-std::unique_ptr<DataLoader<Dataset, Sampler>> make_data_loader(
-    Dataset dataset,
-    DataLoaderOptions options,
-    Sampler sampler) {
-  return torch::make_unique<DataLoader<Dataset, Sampler>>(
+torch::disable_if_t<
+    Dataset::is_stateful,
+    std::unique_ptr<StatelessDataLoader<Dataset, Sampler>>>
+make_data_loader(Dataset dataset, DataLoaderOptions options, Sampler sampler) {
+  return torch::make_unique<StatelessDataLoader<Dataset, Sampler>>(
       std::move(dataset), std::move(options), std::move(sampler));
 }
 
+template <typename Dataset, typename = torch::enable_if_t<Dataset::is_stateful>>
```

Comment:
It would be nice if we could use the same technique to enable/disable overloads. I always feel concerned when I start mixing them...